src/nvim/undo.c

```diff
@@ -2317,74 +2317,77 @@ static void u_undoredo(int undo, bool do_buf_event)
     buf_updates_changedtick(curbuf);
   }
 
-  /*
-   * restore marks from before undo/redo
-   */
-  for (i = 0; i < NMARKS; ++i) {
-    if (curhead->uh_namedm[i].mark.lnum != 0) {
-      free_fmark(curbuf->b_namedm[i]);
-      curbuf->b_namedm[i] = curhead->uh_namedm[i];
+  if (curbuf->b_u_curhead != NULL) {
```

Comment:
📝 add a null guard and fix clint errors.

<details>
<summary>ASan log if no guard</summary>

```diff
diff --git a/src/nvim/undo.c b/src/nvim/undo.c
index d89c8f25e..1b7e3664b 100644
--- a/src/nvim/undo.c
+++ b/src/nvim/undo.c
@@ -2317,7 +2317,7 @@ static void u_undoredo(int undo, bool do_buf_event)
     buf_updates_changedtick(curbuf);
   }
 
-  if (curbuf->b_u_curhead != NULL) {
+  // if (curbuf->b_u_curhead != NULL) {
     // restore marks from before undo/redo
     for (i = 0; i < NMARKS; i++) {
       if (curhead->uh_namedm[i].mark.lnum != 0) {
@@ -2386,7 +2386,7 @@ static void u_undoredo(int undo, bool do_buf_event)
     // The timestamp can be the same for multiple changes, just use the one of
     // the undone/redone change.
     curbuf->b_u_time_cur = curhead->uh_time;
-  }
+  // }
 
   unblock_autocmds();
 #ifdef U_DEBUG
```

```
-- Output to stderr:
=================================================================
==6171==ERROR: AddressSanitizer: heap-use-after-free on address 0x61a000005ae0 at pc 0x000001bfdca3 bp 0x7ffdc1a0ee10 sp 0x7ffdc1a0ee08
READ of size 8 at 0x61a000005ae0 thread T0
    #0 0x1bfdca2 in u_undoredo /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:2323:38
    #1 0x1bed863 in u_doit /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:1774:7
    #2 0x1becad7 in u_undo /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:1672:3
    #3 0xd31401 in ex_undo /home/notomo/workspace/memo/neovim/build/../src/nvim/ex_docmd.c:7897:5
    #4 0xca68a8 in do_one_cmd /home/notomo/workspace/memo/neovim/build/../src/nvim/ex_docmd.c:1961:5
    #5 0xc8ec05 in do_cmdline /home/notomo/workspace/memo/neovim/build/../src/nvim/ex_docmd.c:601:20
    #6 0xc94ef5 in do_cmdline_cmd /home/notomo/workspace/memo/neovim/build/../src/nvim/ex_docmd.c:287:10
    #7 0x6ee755 in nvim_command /home/notomo/workspace/memo/neovim/build/../src/nvim/api/vim.c:147:3
    #8 0x5f6b09 in handle_nvim_command /home/notomo/workspace/memo/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:2881:3
    #9 0x12bd11c in request_event /home/notomo/workspace/memo/neovim/build/../src/nvim/msgpack_rpc/channel.c:380:19
    #10 0xba94ae in multiqueue_process_events /home/notomo/workspace/memo/neovim/build/../src/nvim/event/multiqueue.c:157:7
    #11 0x1378e3c in nv_event /home/notomo/workspace/memo/neovim/build/../src/nvim/normal.c:8079:3
    #12 0x132a53e in normal_execute /home/notomo/workspace/memo/neovim/build/../src/nvim/normal.c:1142:3
    #13 0x1a0d8d1 in state_enter /home/notomo/workspace/memo/neovim/build/../src/nvim/state.c:69:26
    #14 0x12e3429 in normal_enter /home/notomo/workspace/memo/neovim/build/../src/nvim/normal.c:463:3
    #15 0x1041904 in main /home/notomo/workspace/memo/neovim/build/../src/nvim/main.c:557:3
    #16 0x7fc78d83ab96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
    #17 0x45b489 in _start (/home/notomo/workspace/memo/neovim/build/bin/nvim+0x45b489)

0x61a000005ae0 is located 96 bytes inside of 1216-byte region [0x61a000005a80,0x61a000005f40)
freed by thread T0 here:
    #0 0x51b170 in __interceptor_free.localalias.0 (/home/notomo/workspace/memo/neovim/build/bin/nvim+0x51b170)
    #1 0x11eb664 in xfree /home/notomo/workspace/memo/neovim/build/../src/nvim/memory.c:119:3
    #2 0x1c0e4b5 in u_freeentries /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:2830:3
    #3 0x1bdc4d6 in u_freeheader /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:2763:3
    #4 0x1bec437 in u_blockfree /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:2935:5
    #5 0x74a287 in buf_freeall /home/notomo/workspace/memo/neovim/build/../src/nvim/buffer.c:740:5
    #6 0x741b46 in close_buffer /home/notomo/workspace/memo/neovim/build/../src/nvim/buffer.c:560:3
    #7 0x74fd20 in do_buffer /home/notomo/workspace/memo/neovim/build/../src/nvim/buffer.c:1287:9
    #8 0x597626 in nvim_buf_delete /home/notomo/workspace/memo/neovim/build/../src/nvim/api/buffer.c:990:16
    #9 0x55e485 in nlua_msgpack_nvim_buf_delete /home/notomo/workspace/memo/neovim/build/src/nvim/auto/msgpack_lua_c_bindings.generated.c:879:3
    #10 0x1eef02a in lj_BC_FUNCC (/home/notomo/workspace/memo/neovim/build/bin/nvim+0x1eef02a)

previously allocated by thread T0 here:
    #0 0x51b340 in malloc (/home/notomo/workspace/memo/neovim/build/bin/nvim+0x51b340)
    #1 0x11eb2d1 in try_malloc /home/notomo/workspace/memo/neovim/build/../src/nvim/memory.c:71:15
    #2 0x11eb4f9 in xmalloc /home/notomo/workspace/memo/neovim/build/../src/nvim/memory.c:105:15
    #3 0x1bd2ca1 in u_savecommon /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:377:13
    #4 0x1bd1cb2 in u_save /home/notomo/workspace/memo/neovim/build/../src/nvim/undo.c:235:10
    #5 0x5903f7 in nvim_buf_set_lines /home/notomo/workspace/memo/neovim/build/../src/nvim/api/buffer.c:515:7
    #6 0x5baf0e in handle_nvim_buf_set_lines /home/notomo/workspace/memo/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:560:3
    #7 0x12bd11c in request_event /home/notomo/workspace/memo/neovim/build/../src/nvim/msgpack_rpc/channel.c:380:19
    #8 0xba94ae in multiqueue_process_events /home/notomo/workspace/memo/neovim/build/../src/nvim/event/multiqueue.c:157:7
    #9 0x1378e3c in nv_event /home/notomo/workspace/memo/neovim/build/../src/nvim/normal.c:8079:3
    #10 0x132a53e in normal_execute /home/notomo/workspace/memo/neovim/build/../src/nvim/normal.c:1142:3
    #11 0x1a0d8d1 in state_enter /home/notomo/workspace/memo/neovim/build/../src/nvim/state.c:69:26
    #12 0x12e3429 in normal_enter /home/notomo/workspace/memo/neovim/build/../src/nvim/normal.c:463:3
    #13 0x1041904 in main /home/notomo/workspace/memo/neovim/build/../src/nvim/main.c:557:3
    #14 0x7fc78d83ab96 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
```

</details>